core/amber/src/main/scala/edu/uci/ics/texera/web/resource/UserDictionaryResource.scala

```diff
@@ -0,0 +1,238 @@
+      req: PostRequest
+  ): Response = {
+    val user = UserResource.getUser(session).orNull
+    if (user == null)
+      Response.status(Response.Status.UNAUTHORIZED).entity(ErrorResponse("Invalid session")).build()
+    else if (req == null || req.key == null || req.value == null) {
+      Response
+        .status(Response.Status.BAD_REQUEST)
+        .entity(ErrorResponse("Malformed JSON payload"))
+        .build()
+    } else if (dictEntryExists(user, req.key)) {
+      userDictionaryDao.update(new UserDictionary(user.getUid, req.key, req.value))
+      Response.ok(ConfirmationResponse("Updated")).build()
+    } else {
+      userDictionaryDao.insert(new UserDictionary(user.getUid, req.key, req.value))
+      Response.ok(ConfirmationResponse("Created")).build()
+    }
+  }
+
+  /**
+    * This method deletes a key-value pair from the current in-session user's dictionary based on
+    * the "key" attribute of the DeleteRequest
+    * @param session HttpSession
+    * @param req DeleteRequest
+    * @return Response, 200 ok - payload: "Deleted",
+    * 401 unauthorized - payload: "Invalid session",
+    * 400 bad request - payload: ErrorResponse("Malformed JSON payload"),
+    * 422 Unprocessable Entity - payload: "no such entry" (if no entry exists for provided key)
+    */
+  @DELETE
+  @Path("/delete")
+  def deleteValue(@Session session: HttpSession, req: DeleteRequest): Response = {
+    val user = UserResource.getUser(session).orNull
+    if (user == null)
+      Response.status(Response.Status.UNAUTHORIZED).entity(ErrorResponse("Invalid session")).build()
+    else if (req == null || req.key == null) {
+      Response
+        .status(Response.Status.BAD_REQUEST)
+        .entity(ErrorResponse("Malformed JSON payload"))
+        .build()
+    } else if (!dictEntryExists(user, req.key)) {
+      Response.status(422).entity(ErrorResponse("No such entry")).build()
+    } else {
+      deleteDictEntry(user, req.key)
+      Response.ok(ConfirmationResponse("Deleted")).build()
+    }
+  }
+
+  /**
+    * This method retrieves a value from the user_dictionary table
+    * given a user's uid and key. each tuple (uid, key) is a primary key
+    * in user_dictionary, and should uniquely identify one value
+    */
+  private def getValueByKey(user: User, key: String): String = {
+    SqlServer.createDSLContext
+      .fetchOne(
+        USER_DICTIONARY,
+        USER_DICTIONARY.UID.eq(user.getUid).and(USER_DICTIONARY.KEY.eq(key))
+      )
+      .getValue
+  }
+
+  /**
+    * This method retrieves all of a user's dictionary entries in
+    * the user_dictionary table as a json object
+    */
+  private def getDict(user: User): Map[String, String] = {
+    Map(
+      asScalaBuffer(
+        SqlServer.createDSLContext
+          .select()
+          .from(USER_DICTIONARY)
+          .where(USER_DICTIONARY.UID.eq(user.getUid))
+          .fetchInto(classOf[UserDictionary])
+      ) map { entry => (entry.getKey, entry.getValue) }: _*
```

Comment:
I am not familiar with this `: _*` operator. My understanding is required return type is a `Map[String, String]` but you are having a `map` iterator. Some google tells me `: _*` is a special type which forces compiler to treat something as a collection. Can we avoid using `: _*` my gut feeling tells me to avoid this kind of special type.

Reply:
If you're familiar with python, it's similar to the * operator that unpacks a list into args. 
```
a =  [1,10]
list(range(*a)) == [1,2,3,4,5,6,7,8,9] 
```
The `:` is a type ascription operation, kind of a type hint for the compiler. What follows should describe the type of the arguments
The `_` is a placeholder that can accept any type/value
The `*` unpacks the list into individual arguments, which are fed into the scala Map constructor

I don't think this is very uncommon in scala.

Reply:
ok

Reply:
`:_*` is kind of like the python `*` operator that unpacks a list into arguments, ex:
```
a = [1,10]
list(range(*a)) == list(range(1,10)) == [1,2,3,4,5,6,7,8,9]
```
the `:` is a *type ascription*, i.e `<value> : <type>` is a compile time hint that `<value>` is of type `<type>`
the `_` is a placeholder, allowing any value/type
the `*` actually unpacks the list into arguments

It doesn't look very intuitive, but it seems to be the prescribed way to pass a list into a varargs scala function.

There are of course other ways to make a Map, but I couldn't get them to work for some crazy reason :(